### PR TITLE
Support Arc experimental

### PR DESCRIPTION
**Pull request by ben-kaufman**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -31,6 +31,6 @@ services:
     environment:
```
Comment (orenyodfat) [inline docker-compose.yml]: - use graph node 0.16.1 https://github.com/graphprotocol/graph-node/releases/tag/v0.16.1
- IPFS max file size should be increased to 300000.

File: ops/generate-daos-subgraph.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ const { migrationFileLocation: defaultMigrationFileLocation, network, startBlock
 const {   subgraphLocation: defaultSubgraphLocation } = require('./graph-cli')
 const path = require("path");
 const currentDir = path.resolve(`${__dirname}`)
+const supportedSchemes = ['ContributionReward', 'GenericScheme', 'SchemeRegistrar']
```
Comment (orenyodfat) [inline ops/generate-daos-subgraph.js]: what is this ?

File: ops/mappings.json [hunk 3]
```diff
@@ -2,623 +2,56 @@
...
-        "mapping": "DAOTracker",
-        "arcVersion": "0.0.1-rc.33"
-      }
-    ]
+        {
+          "name": "GEN",
+          "contractName": "GEN",
+          "dao": "package",
+          "mapping": "DAOToken",
+          "arcVersion": "0.0.1-rc.2"
```
Comment (orenyodfat) [inline ops/mappings.json]: this arc version might be confusing with the legacy one .
maybe we should give it other name ? 
`1.0.1-rc.2` or `exp-0.0.1-rc.2`?

File: ops/utils.js [hunk 4]
```diff
@@ -2,7 +2,7 @@ const fs = require("fs");
 
 function versionToNum(version) {
   const strlen = version.length;
-  return Number(version.slice(strlen - 2, strlen));
+  return Number(version.slice(strlen - 1, strlen));
```
Comment (orenyodfat) [inline ops/utils.js]: how about checking the Len of the version first and then decide if to do strlen-1 or strlen-2 ? this will works for both 2 and 1 digits rc versions.

File: src/domain/dao.ts [hunk 5]
```diff
@@ -42,23 +42,11 @@ export function insertNewDAO(
   dao.nativeToken = nativeTokenAddress.toHex();
   dao.nativeReputation = nativeReputationAddress.toHex();
   dao.reputationHoldersCount = BigInt.fromI32(0);
-  dao.register = 'na';
   saveDAO(dao);
 
   return dao;
 }
 
-export function register(
```
Comment (orenyodfat) [inline src/domain/dao.ts]: why is this removed?

File: src/domain/schema.graphql [hunk 6]
```diff
@@ -29,7 +29,6 @@ type DAO @entity {
 	reputationHolders: [ReputationHolder!] @derivedFrom(field: "dao")
 	reputationHoldersCount: BigInt!
 	rewards: [GPReward!] @derivedFrom(field: "dao")
-	register: String!
```
Comment (orenyodfat) [inline src/domain/schema.graphql]: why?

File: src/mappings/DAOTracker/mapping.ts [hunk 7]
```diff
@@ -12,8 +12,8 @@ import {
   AvatarContract,
```
Comment (orenyodfat) [inline src/mappings/DAOTracker/mapping.ts]: doa tracker should be removed.
we should track new daos  according to the daoFactory events.

**Comment by orenyodfat** [comment]
increment package version to "exp-0.1.0-1",

File: ops/generate-daos-subgraph.js (hunk 2 above)
Comment (ben-kaufman) [reply ops/generate-daos-subgraph.js]: To index properly the schemes which the subgraph support and implemented.

File: ops/mappings.json (hunk 3 above)
Comment (ben-kaufman) [reply ops/mappings.json]: Ok, shouldn't be a problem, but I'll need you to first rename the arc package like this.

File: src/domain/dao.ts (hunk 5 above)
Comment (ben-kaufman) [reply src/domain/dao.ts]: Beacuse we don't have DAORegistry with the new Arc.

File: src/domain/schema.graphql (hunk 6 above)
Comment (ben-kaufman) [reply src/domain/schema.graphql]: Because we removed DAORegistry from Arc.

File: src/mappings/DAOTracker/mapping.ts (hunk 7 above)
Comment (ben-kaufman) [reply src/mappings/DAOTracker/mapping.ts]: But DAOTracker is part of DAOFactory, DAOFactory uses it and accept it as an initialize parameter.

File: ops/generate-daos-subgraph.js (hunk 2 above)
Comment (orenyodfat) [reply ops/generate-daos-subgraph.js]: why it is needed to set explicitly here. ? 
how about daoFactory ?

Comment (ben-kaufman) [reply ops/generate-daos-subgraph.js]: How is that related to the DAOFactory? This is related to which schemes the subgraph has mappings for and is indexing with full support, namely, the GenericScheme, ContributionReward, and SchemeRegistrar. It is not related to Arc or the DAOFactory in any way.

Comment (orenyodfat) [reply ops/generate-daos-subgraph.js]: why it is needed to set explicitly here. ?

File: src/domain/dao.ts (hunk 5 above)
Comment (orenyodfat) [reply src/domain/dao.ts]: DAORegistry was never part of arc . it comes from arc-hive repo https://github.com/daostack/arc-hive/blob/master/contracts/DAORegistry.sol .
we need that.

File: src/domain/schema.graphql (hunk 6 above)
Comment (orenyodfat) [reply src/domain/schema.graphql]: DAORegistry was never part of arc . it comes from arc-hive repo https://github.com/daostack/arc-hive/blob/master/contracts/DAORegistry.sol .
we need that.

File: src/mappings/DAOTracker/mapping.ts (hunk 7 above)
Comment (orenyodfat) [reply src/mappings/DAOTracker/mapping.ts]: will remove it  from daoFactory

File: ops/generate-daos-subgraph.js (hunk 2 above)
Comment (ben-kaufman) [reply ops/generate-daos-subgraph.js]: So that we can know if we should generate entities for them or not.

File: src/domain/dao.ts (hunk 5 above)
Comment (ben-kaufman) [reply src/domain/dao.ts]: Oh ok. I thought it will be removed with the new Architecture.

File: src/mappings/DAOFactory/mapping.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+} from '../../contractsInfo';
+
+import { Avatar } from '../../types/Controller/Avatar';
+import { DAOFactory, InitialSchemesSet, NewOrg, ProxyCreated, SchemeInstance } from '../../types/DAOFactory/DAOFactory';
+import {
+  AvatarContract, ContractInfo, DAOFactoryContract,
+} from '../../types/schema';
+import { createTemplate, debug, fetchTemplateName, setContractInfo } from '../../utils';
+
+export function getDAOFactoryContract(address: Address): DAOFactoryContract {
```
Comment (orenyodfat) [inline src/mappings/DAOFactory/mapping.ts]: is this must be export function ?

File: src/mappings/DAOFactory/mapping.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    daoFactory.app = daoFactoryContract.app();
+    daoFactory.save();
+    setContractsInfo();
+    setTemplatesInfo();
+  }
+  return daoFactory;
+}
+
+export function handleNewOrg(event: NewOrg): void {
+  // Ensure the DAOTrackerContract has been added to the store
```
Comment (orenyodfat) [inline src/mappings/DAOFactory/mapping.ts]: DAOFactory

File: src/mappings/DAOFactory/mapping.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  getDAOFactoryContract(event.address);
+
+  let avatar = event.params._avatar;
+
+  // If the avatar already exists, early out
+  if (AvatarContract.load(avatar.toHex()) != null) {
+    return;
+  }
+
+  let avatarContract = Avatar.bind(avatar);
```
Comment (orenyodfat) [inline src/mappings/DAOFactory/mapping.ts]: does arc/package version is ignored here ?

File: src/mappings/DAOFactory/mapping.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  let arcVersion: string;
+
+  let avatarInfo = ContractInfo.load(avatar.toHex());
+  if (avatarInfo != null) {
+    avatarInfo.name = 'Avatar';
+    avatarInfo.alias = 'Avatar';
+    avatarInfo.save();
+    arcVersion = avatarInfo.version;
+  } else {
+    // If contractInfo doesn't exist something went wrong and should exit early
```
Comment (orenyodfat) [inline src/mappings/DAOFactory/mapping.ts]: is that won't be the case on any new org created via dao factory ?

Comment (ben-kaufman) [reply src/mappings/DAOFactory/mapping.ts]: Not sure I understand the question, but the `ProxyCreated` event is fired before the NewOrg and thus make sure the contractInfo exists beforehand. This check is just to make sure the subgraph won't crash if somehow the contractInfo doesn't exist (though it always should).

File: src/mappings/DAOFactory/mapping.ts (hunk 10 above)
Comment (ben-kaufman) [reply src/mappings/DAOFactory/mapping.ts]: Basically yes. We did not yet decide how we should deal with multi version support in the new subgraph so for now we still have this "version dependency".

File: src/mappings/DAOFactory/mapping.ts (hunk 11 above)
Comment (orenyodfat) [reply src/mappings/DAOFactory/mapping.ts]: contractInfo is a preCompiled data. 
it does not hold avatar addresses.

Comment (orenyodfat) [reply src/mappings/DAOFactory/mapping.ts]: is there a test for this mapping ? and specific for this functionality ?

Comment (ben-kaufman) [reply src/mappings/DAOFactory/mapping.ts]: It can be precompiled, but not necessarily only so, and it also had `Avatar` and all other contracts. There's no reason not to use it here now that we only auto index daos.

There is a short test for that. https://github.com/daostack/subgraph/blob/subgraph-experimental/test/0.1.1-rc.1/contractsInfo.spec.ts

We could add more though doesn't seem necessary as this is a very simple mapping.
